src/terrain/generation/biome.cpp

```diff
@@ -98,6 +135,108 @@ Biome::Biome(const std::string& biome_name) {
+    // add tile macro to tiles
+    for (const Json::Value& tile_type : biome_data["Tile_Data"]) {
+        std::vector<TileMacro_t> tile_macros;
+        for (const Json::Value& tile_macro_id : tile_type["Land_From"]) {
+            TileMacro_t tile_macro = tile_macro_id.asInt();
+            if (tile_macro >= land_generators_.size()) [[unlikely]] {
+                LOG_WARNING(
+                    logging::terrain_logger,
+                    "Tile Macro {} is not valid. There are only {} land generators.",
+                    tile_macro, land_generators_.size()
+                );
+                continue;
+            }
+            tile_macros.push_back(tile_macro_id.asInt());
+        }
+        macro_tile_types_.push_back(std::move(tile_macros));
+    }
+}
+
+void Biome::read_add_to_top_data(const Json::Value& after_effects_data) {
+    for (const Json::Value& add_data : after_effects_data["Add_To_Top"]){
+        AddToTop generator(add_data);
+        add_to_top_generators_.push_back(std::move(generator));
+    }
+}
+
+std::map<MaterialId, const terrain::Material>
+Biome::init_materials(const Json::Value& material_data) {
+    std::map<MaterialId, const terrain::Material> out;
+    for (auto element_it = material_data.begin(); element_it != material_data.end();
+         element_it++) {
+        std::vector<std::pair<const std::string, ColorInt>> color_vector;
+
+        const Json::Value material = *element_it;
+        std::string name = element_it.key().asString();
+        for (const Json::Value& json_color : material["colors"]) {
+            const std::string color_name = json_color["name"].asString();
+            ColorInt color_value = std::stoll(json_color["hex"].asString(), 0, 16);
+            color_vector.push_back(std::make_pair(std::move(color_name), color_value));
+        }
+
+        terrain::Material mat{
+            color_vector,                                    // color
+            static_cast<float>(material["speed"].asFloat()), // speed_multiplier
+            material["solid"].asBool(),                      // solid
+            static_cast<MaterialId>(material["id"].asInt()), // element_id
+            name
+        }; // name
+        out.insert(std::make_pair(mat.element_id, mat));
+    }
+    // I think this should be delayed
+    //terrain::TerrainColorMapping::assign_color_mapping(out);
+    return out;
+}
+
+std::map<ColorInt, std::pair<const Material*, ColorId>>
+Biome::get_colors_inverse_map() const {
+    std::map<ColorInt, std::pair<const Material*, ColorId>> materials_inverse;
+    for (auto it = materials_.begin(); it != materials_.end(); it++) {
+        for (size_t color_id = 0; color_id < it->second.color.size(); color_id++) {
+            materials_inverse.insert(
+                std::map<ColorInt, std::pair<const Material*, ColorId>>::value_type(
+                    it->second.color.at(color_id).second,
+                    std::make_pair(&it->second, (ColorId)color_id)
```

Comment:
no c-style casts

and don't take pointers inside collections as the collections may change, and then all ptrs will be invaliadated